### PR TITLE
feature/83: Improved performance of loading many small files

### LINKED ISSUE #83: Improve performance of loading many small files
Improve performance of loading many small files

### PR DESCRIPTION
**Pull request by jakobbraun**
Closes #83

### CODE REVIEW

File: src/main/java/com/exasol/adapter/document/documentfetcher/files/RemoteFilePrefetchingIterator.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        this.next = nextCandidate.get();
+    }
+
+    private void wait10Ms() {
+        try {
+            Thread.sleep(10);
+        } catch (final InterruptedException exception) {
+            Thread.currentThread().interrupt();
+            throw new IllegalStateException(ExaError.messageBuilder("F-VSDF-21")
+                    .message("Interrupted while waiting for prefetchable.").toString(), exception);
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/documentfetcher/files/RemoteFilePrefetchingIterator.java:65]:
```suggestion
                    .message("Interrupted while waiting for buffered ready or non-prefetchable file.").toString(), exception);
```

File: src/main/java/com/exasol/adapter/document/documentfetcher/files/TooManyRequestsException.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package com.exasol.adapter.document.documentfetcher.files;
+
+/**
+ * Exception thrown by the iterator returned by {@link FileLoader} if the source throttled the request.
+ */
+public class TooManyRequestsException extends RuntimeException {
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/documentfetcher/files/TooManyRequestsException.java:6]: This is only thrown from tests. Should this be the one from s3 sdk?

File: doc/changes/changes_4.0.0.md [hunk 3]
```diff
@@ -2,6 +2,10 @@
 
 Code name:
 
+## Features:
```
Comment (morazow) [inline doc/changes/changes_4.0.0.md:5]:
```suggestion
## Features
```

**Review by morazow — CHANGES_REQUESTED** [review]
Looks good, added some suggestions and questions.

File: src/main/java/com/exasol/adapter/document/documentfetcher/files/RemoteFilePrefetchingIterator.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+                    this.buffer.remove(each);
+                    this.dynamicPrefetchSize += 0.1;
+                    if (this.dynamicPrefetchSize > MAX_PREFETCH) {
+                        this.dynamicPrefetchSize = MAX_PREFETCH;
+                    }
+                    return Optional.of(loadedFile);
+                } catch (final TooManyRequestsException exception) {
+                    this.retryQueue.add(each);
+                    this.buffer.remove(each);
+                    this.dynamicPrefetchSize = this.buffer.size() - 1f;
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/document/documentfetcher/files/RemoteFilePrefetchingIterator.java:83]: exactly.

File: src/main/java/com/exasol/adapter/document/documentfetcher/files/TooManyRequestsException.java (hunk 2 above)
Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/document/documentfetcher/files/TooManyRequestsException.java:6]: it's thrown in the implementation of the s3-virtual-schema

File: src/main/java/com/exasol/adapter/document/documentfetcher/files/RemoteFilePrefetchingIterator.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+            if (each.isDone()) {
+                try {
+                    final RemoteFile loadedFile = each.getLoadedFile();
+                    this.buffer.remove(each);
+                    this.dynamicPrefetchSize += 0.1;
+                    if (this.dynamicPrefetchSize > MAX_PREFETCH) {
+                        this.dynamicPrefetchSize = MAX_PREFETCH;
+                    }
+                    return Optional.of(loadedFile);
+                } catch (final TooManyRequestsException exception) {
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/document/documentfetcher/files/RemoteFilePrefetchingIterator.java:80]: each.getLoadedFile(); can throw if the loading failed due to rate limits

File: src/main/java/com/exasol/adapter/document/documentfetcher/files/RemoteFilePrefetchingIterator.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+                    .message("Interrupted while waiting for prefetchable.").toString(), exception);
+        }
+    }
+
+    private Optional<RemoteFile> findReady() {
+        for (final LoadingRemoteFile each : new ArrayList<>(this.buffer)) {
+            if (each.isDone()) {
+                try {
+                    final RemoteFile loadedFile = each.getLoadedFile();
+                    this.buffer.remove(each);
```
[the remark this answers is not in the document]
Comment (jakobbraun) [reply src/main/java/com/exasol/adapter/document/documentfetcher/files/RemoteFilePrefetchingIterator.java:74]: I refactored to use iterator

**Review by morazow — APPROVED** [review]

**Pull request merged**